File: ts/src/tracker.test.ts
import {Tracker} from "./tracker";
import {Backend} from "./backend";
import {Account, Application, Intent, Interaction, InteractionChannel, Stage, Subject} from "./generated/tracking";

function mockBackend(): Backend {
    return {
        track: jest.fn(),
    }
}

test("stage event tracking", async () => {

    const backend = mockBackend();
    const tracker = new Tracker(backend);

    const eventName = "submitted.meter-reading-submit";

    const account: Account = {number: "acc-123", id: ""};
    const application: Application = {id: "acc-123"};
    const subject: Subject = Subject.SUBJECT_METER_READING;
    const intent: Intent = Intent.INTENT_METER_READING_SUBMIT;
    const stage: Stage = Stage.STAGE_SUBMITTED;
    const attributes = {"a-a": "b", c: 'd'};
    const expectedAttributes = {"a_a": "b", c: 'd'};

    await tracker.trackStage(
        account,
        application,
        subject,
        intent,
        stage,
        new Promise(r => r(attributes)),
    )

    expect(backend.track).toBeCalledWith(eventName, {
        "account_number": account.number,
        client_id: application.id,
        subject: "meter-reading",
        intent: "meter-reading-submit",
        stage: "submitted",
        ...expectedAttributes
    });
})

test("interaction event tracking", async () => {

    const backend = mockBackend();
    const tracker = new Tracker(backend);

    const eventName = "email.clicked";

    const account: Account = {number: "acc-123", id: ""};
    const application: Application = {id: "acc-123"};
    const subject: Subject = Subject.SUBJECT_METER_READING;
    const intent: Intent = Intent.INTENT_METER_READING_SUBMIT;
    const interaction: Interaction = Interaction.INTERACTION_CLICKED;
    const channel: InteractionChannel = InteractionChannel.INTERACTION_CHANNEL_EMAIL;
    const attributes = {"a-a": "b", c: 'd'};
    const expectedAttributes = {"a_a": "b", c: 'd'};

    await tracker.trackInteraction(
        account,
        application,
        subject,
        intent,
        interaction,
        channel,
        new Promise(r => r(attributes)),
    )

    expect(backend.track).toBeCalledWith(eventName, {
        "account_number": account.number,
        client_id: application.id,
        subject: "meter-reading",
        intent: "meter-reading-submit",
        interaction: "clicked",
        interaction_channel: "email",
        ...expectedAttributes
    });
})

test("visit event tracking", async () => {

    const backend = mockBackend();
    const tracker = new Tracker(backend);

    const eventName = "visit";

    const account: Account = {number: "acc-123", id: ""};
    const application: Application = {id: "acc-123"};
    const location = "location";
    const attributes = {"a-a": "b", c: 'd'};
    const expectedAttributes = {"a_a": "b", c: 'd'};

    await tracker.trackVisit(
        account,
        application,
        location,
        new Promise(r => r(attributes)),
    )

    expect(backend.track).toBeCalledWith(eventName, {
        "account_number": account.number, client_id: application.id, location, ...expectedAttributes
    });
})

test("click event tracking", async () => {

    const backend = mockBackend();
    const tracker = new Tracker(backend);

    const eventName = "click";

    const account: Account = {number: "acc-123", id: ""};
    const application: Application = {id: "acc-123"};
    const target = "target";
    const attributes = {"a-a": "b", c: 'd'};
    const expectedAttributes = {"a_a": "b", c: 'd'};

    await tracker.trackClick(
        account,
        application,
        target,
        new Promise(r => r(attributes)),
    )

    expect(backend.track).toBeCalledWith(eventName, {
        "account_number": account.number, client_id: application.id, target, ...expectedAttributes
    });
})

File: go/tracking/tracker_test.go
package tracking_test

import (
	"context"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/utilitywarehouse/customer-tracking-sdk/go/tracking"
	"github.com/utilitywarehouse/customer-tracking-sdk/go/types"
)

func TestTrackStage(t *testing.T) {
	ctx := context.Background()
	e := &types.StageEvent{
		Actor: &types.Actor{
			Id:         "000000",
			Attributes: map[string]string{"account-number": "000001"},
		},
		Application: &types.Application{
			Id: "test-application",
		},
		Subject: types.SUBJECT_METER_READING,
		Intent:  types.INTENT_METER_READING_SUBMIT,
		Stage:   types.STAGE_COMPLETED,
		Attributes: map[string]string{
			"sample key": "foo",
		},
	}

	b := &mockBackend{}
	tracker := tracking.New(b)
	err := tracker.TrackStage(ctx, e)
	require.NoError(t, err)

	require.Len(t, b.events, 1)
	receivedEvent := b.events[0]

	assert.Equal(t, "completed.meter-reading-submit", receivedEvent.name)
	assert.Equal(t, "000000", receivedEvent.id)
	expectedAttributes := map[string]string{
		"account_number": "000001",
		"client_id":      "test-application",
		"subject":        "meter-reading",
		"intent":         "meter-reading-submit",
		"stage":          "completed",
		"sample_key":     "foo",
	}
	assert.Equal(t, expectedAttributes, receivedEvent.attributes)
}

func TestTrackInteraction(t *testing.T) {
	ctx := context.Background()
	e := &types.InteractionEvent{
		Actor: &types.Actor{
			Id:         "000000",
			Attributes: map[string]string{"account-number": "000001"},
		},
		Application: &types.Application{
			Id: "test-application",
		},
		Subject:     types.SUBJECT_METER_READING,
		Intent:      types.INTENT_METER_READING_SUBMIT,
		Channel:     types.INTERACTION_CHANNEL_EMAIL,
		Interaction: types.INTERACTION_VIEWED,
		Attributes: map[string]string{
			"sample key":  "foo",
			"sample key2": "foo",
		},
	}

	b := &mockBackend{}
	tracker := tracking.New(b)
	err := tracker.TrackInteraction(ctx, e)
	require.NoError(t, err)

	require.Len(t, b.events, 1)
	receivedEvent := b.events[0]

	assert.Equal(t, "email.viewed", receivedEvent.name)
	assert.Equal(t, "000000", receivedEvent.id)
	expectedAttributes := map[string]string{
		"account_number":      "000001",
		"client_id":           "test-application",
		"subject":             "meter-reading",
		"intent":              "meter-reading-submit",
		"interaction_channel": "email",
		"interaction":         "viewed",
		"sample_key":          "foo",
		"sample_key2":         "foo",
	}
	assert.Equal(t, expectedAttributes, receivedEvent.attributes)
}

type event struct {
	name       string
	id         string
	attributes map[string]string
}

type mockBackend struct {
	events []event
}

func (b *mockBackend) Import(_ context.Context, name string, id string, attrs map[string]string) error {
	return nil
}

func (b *mockBackend) Track(_ context.Context, name string, id string, attrs map[string]string) error {
	e := event{
		name:       name,
		id:         id,
		attributes: attrs,
	}
	b.events = append(b.events, e)
	return nil
}

func (b *mockBackend) Alias(ctx context.Context, currentID string, alias string) error {
	return nil
}

func (b *mockBackend) Close() error {
	return nil
}
